src/section/shared_state.rs

```diff
@@ -52,25 +53,34 @@ impl SharedState {
         }
     }
 
-    pub fn update(&mut self, new: Self) {
+    pub fn update(&mut self, new: Self) -> Result<(), RoutingError> {
         if self.handled_genesis_event {
-            log_or_panic!(
-                log::Level::Error,
-                "shared state update - genesis event already handled",
-            );
+            error!("shared state update - genesis event already handled",);
+            return Err(RoutingError::InvalidState);
         }
 
-        if self.our_history.len() > 1 && *self != new {
-            log_or_panic!(
-                log::Level::Error,
-                "shared state update - mismatch: old: {:?} --- new: {:?}",
-                self,
-                new
-            );
+        if self.our_history.len() > 1 {
+            if *self != new {
+                error!(
+                    "shared state update - mismatch: old: {:?} --- new: {:?}",
+                    self, new
+                );
+                return Err(RoutingError::InvalidState);
+            }
+        } else if !new.self_verify() {
```

Comment:
One more thing: verify that `self.our_history` and `new.our_history` have at least one key in common. Without it someone could in theory send us a completely fabricated `SharedState` which would still pass `self_verify`.

Reply:
ok. will do